server/sonar-server/src/main/java/org/sonar/server/authentication/JwtSerializer.java

```diff
@@ -0,0 +1,178 @@
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+import org.sonar.api.Startable;
+import org.sonar.api.config.Settings;
+import org.sonar.api.server.ServerSide;
+import org.sonar.api.utils.System2;
+import org.sonar.core.util.UuidFactory;
+import org.sonar.server.exceptions.ForbiddenException;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.jsonwebtoken.impl.crypto.MacProvider.generateKey;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class can be used to encode or decode a JWT token
+ */
+@ServerSide
+public class JwtSerializer implements Startable {
+
+  private static final String SECRET_KEY_PROPERTY = "sonar.jwt.base64hs256secretKey";
+  private static final String ISSUER = "sonarqube";
+
+  private static final SignatureAlgorithm SIGNATURE_ALGORITHM = SignatureAlgorithm.HS256;
+
+  private final Settings settings;
+  private final System2 system2;
+  private final UuidFactory uuidFactory;
+
+  private SecretKey secretKey;
+
+  public JwtSerializer(Settings settings, System2 system2, UuidFactory uuidFactory) {
+    this.settings = settings;
+    this.system2 = system2;
+    this.uuidFactory = uuidFactory;
+  }
+
+  @Override
+  public void start() {
+    String encodedKey = settings.getString(SECRET_KEY_PROPERTY);
+    if (encodedKey == null) {
+      SecretKey newSecretKey = generateSecretKey();
+      settings.setProperty(SECRET_KEY_PROPERTY, Base64.getEncoder().encodeToString(newSecretKey.getEncoded()));
+      this.secretKey = newSecretKey;
+    } else {
+      this.secretKey = decodeSecretKeyProperty(encodedKey);
+    }
+  }
+
+  String encode(JwtSession jwtSession) {
+    checkIsStarted();
+    long now = system2.now();
+    JwtBuilder jwtBuilder = Jwts.builder()
+      .setId(uuidFactory.create())
+      .setSubject(jwtSession.getUserLogin())
+      .setIssuer(ISSUER)
+      .setIssuedAt(new Date(now))
+      .setExpiration(new Date(now + jwtSession.getExpirationTimeInSeconds() * 1000))
+      .signWith(SIGNATURE_ALGORITHM, secretKey);
+    for (Map.Entry<String, Object> entry : jwtSession.getProperties().entrySet()) {
+      jwtBuilder.claim(entry.getKey(), entry.getValue());
+    }
+    return jwtBuilder.compact();
+  }
+
+  Optional<Claims> decode(String token) {
+    checkIsStarted();
+    try {
+      Claims claims = Jwts.parser()
+        .requireIssuer(ISSUER)
+        .setSigningKey(secretKey)
+        .parseClaimsJws(token)
+        .getBody();
+      requireNonNull(claims.getId(), "Token id hasn't been found");
+      requireNonNull(claims.getSubject(), "Token subject hasn't been found");
+      requireNonNull(claims.getExpiration(), "Token expiration date hasn't been found");
+      requireNonNull(claims.getIssuedAt(), "Token creation date hasn't been found");
+      return Optional.of(claims);
+    } catch (ExpiredJwtException | SignatureException e) {
+      return Optional.empty();
+    } catch (Exception e) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)